### PR TITLE
*: Refactor datum arithmetic codes and support Add/Div in xeval.

### PR DESCRIPTION
**Pull request by shenli**
1. Move some functions from evaluator package to types package.
2. Support add, div in xapi.

@coocood @BusyJay

### CODE REVIEW

File: executor/executor_xapi.go [hunk 1]
```diff
@@ -907,6 +907,8 @@ func (b *executorBuilder) binopToPBExpr(client kv.Client, expr *ast.BinaryOperat
 		tp = tipb.ExprType_And
 	case opcode.OrOr:
 		tp = tipb.ExprType_Or
+	case opcode.Plus:
+		tp = tipb.ExprType_Plus
```
Comment (coocood) [inline executor/executor_xapi.go]: Missing `opcode.Div`?

Comment (shenli) [inline executor/executor_xapi.go]: PTAL @coocood

**Comment by zimulala** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

File: xapi/xeval/eval.go [hunk 2]
```diff
@@ -544,3 +546,33 @@ func (e *Evaluator) decodeValueList(valueListExpr *tipb.Expr) (*decodedValueList
...
+		return result, errors.Trace(err)
+	}
+
+	b, err := types.CoerceArithmetic(right)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	a, b = types.CoerceDatum(a, b)
+	if a.IsNull() || b.IsNull() {
+		return result, errors.Trace(err)
```
Comment (BusyJay) [inline xapi/xeval/eval.go]: Seems no need to trace.

**Comment by shenli** [comment]
@BusyJay PTAL

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**